Further updates to msh plus

WPringle opened the pull request:
Major change to allow msh plus to correctly preserve weir triangulation and carry over the nodestrings
- I put the collapse thin triangles call inside clean which now comes after delete bad boundary elements
- I let the clean also handle the smoothing so I took this out of the plus
- Clean also handles the iterations if the quality becomes negative after smoothing so no longer need the while loop in msh.plus
- setProj was copied from dev. This allows one to call setProj to add the desired projection into the msh class
- Also reformatted dpoly to improve the memory management (the inpoly is now also subsetted).

WPringle committed:
Further updates to msh plus

Major change to allow msh plus to correctly preserve weir triangulation and carry over the nodestrings
- I put the collapse thin triangles call inside clean whicn now comes after delete bad boundary elements
- I let the clean also handle the smoothing so I took this out of the plus
- Clean also handles the iterations if the quality becomes negative after smoothing so no longer need the while loop in msh.plus
- setProj was copied from dev. This allows one to call setProj to add the desired projection into the msh class
- Also reformatted dpoly to improve the memory management (the inpoly is now also subsetted).

WPringle committed:
Update meshgen.m

addition to make sure the line is correctly removed for the global domain

WPringle committed:
cirtical fix for global meshing

- Major fix is to the ideal_bars by making sure we get the mid-point in projected coordinates before converting back to lat-lon.
- also I made sure that edgefx was cyclical so that -180 == -180 when queried.
- I increased some default parameters for clean to be stricter.

WPringle commented:
Also critical fix for global meshing

- Major fix is to the ideal_bars by making sure we get the mid-point in projected coordinates before converting back to lat-lon.
- I made sure that edgefx was cyclical so that -180 == 180 when queried.
- I increased some default parameters for clean to be stricter.

WPringle committed:
adding help to msh.interp

krober10nd commented on @msh/msh.m:
```diff
@@ -1572,9 +1576,44 @@ function write(obj,fname,type)
+            end
+
+            % checking for weirs in obj1 to keep
+            if any(obj1.bd.ibtype == 24)
```
This is problematic because a mesh may or may not have its bd structure populated. If it isn't populated this will incur a segfault.

WPringle replied:
True. This has been fixed in 7th commit.

krober10nd commented on @msh/msh.m:
```diff
@@ -956,6 +956,9 @@ function write(obj,fname,type)
                 end
+                % kjr, collapse small triangles together 
+                [obj.p,obj.t] = ...
+                    collapse_thin_triangles(obj.p,obj.t,opt.db); 
```
Note that when you execute collapse, this can alter the node numbering and thus affect the mapping of the bathymetry data to the cleaned mesh later on. In the examples I've tried, after cleaning outside of the meshgen (i.e., after bathy is interpolated) leads to a mess with the bathymetry

krober10nd replied:
Also, if you build a floodplain onto a mesh with a weir, the collapse thin will decimate the weir. You can try by adding a weir to the Houston GBAY example  with floodplain.

krober10nd replied:
It seems that if you are meshing with a weir, it's best to use a custom configuration for clean. For some reason lowering the djc to 0.01 from 0.25 kept the weir in tact. When it was kept on, the clean destroyed the weir even though it was being constrained during the mesh generation.

WPringle replied:
The node numbering was affected even without the collapse routine. We've always had a problem with bathymetry getting screwed during cleaning. So we always have just told people (or just ourselves) to only interp bathymetry on after cleaning. Alternatively we need a mapping routine in there.

WPringle replied:
> It seems that if you are meshing with a weir, it's best to use a custom configuration for clean. For some reason lowering the djc to 0.01 from 0.25 kept the weir in tact. When it was kept on, the clean destroyed the weir even though it was being constrained during the mesh generation.

Yes, you can set djc to 0. When cleaning with a weir I would use the passive option which uses djc = 0. So I would probably turn off the cleanup during mesh generation and then clean manually after the build. Alternatively, we can change the cleanup option to be varargins for the clean call after build.

WPringle replied:
> Note that when you execute collapse, this can alter the node numbering and thus affect

The node numbering was affected even without the collapse routine. We've always had a problem with bathymetry getting screwed during cleaning. So we always have just told people (or just ourselves) to only interp bathymetry on after cleaning. Alternatively we need a mapping routine in there.

krober10nd replied:
Yea, I had forgotten that happens. I'll create an issue.

krober10nd commented on @meshgen/meshgen.m:
```diff
@@ -782,16 +784,17 @@
-                ix(1:nfix)=0;
-                if sum(ix) > 0
-                    dgradx = (feval(obj.fd,[p(ix,1)+deps,p(ix,2)],obj,[])...%,1)...
+                ix(1:nfix) = 0;
```
In general, if the points aren't projecting nicely back into the domain, one simple solution is to execute the project points back twice per meshing iteration.

WPringle replied:
It is working fine.

krober10nd commented:
For the following files (one inset with a weir and regional mesh without it) the new changes do not work. Please try it out. I had to roll back to before your commit to get it to work (only merging from .mats). 
[MeshesForTestingMerge.zip](https://github.com/CHLNDDEV/OceanMesh2D/files/3907773/MeshesForTestingMerge.zip)

krober10nd commented:
Here's the .mat file for the inset with the weir. 
[FireIslandInlet.mat.zip](https://github.com/CHLNDDEV/OceanMesh2D/files/3907775/FireIslandInlet.mat.zip)

krober10nd commented:
And here is the rest of the insets plus the regional GSBv4 mesh. There's three insets and they should all automatically merge into the regional GSBv4 mesh without changing anything outside of their respective insets. This is sufficiently complex to test this pull request. We shouldn't merge anything until this works. 
[TestMergerComplete.zip](https://github.com/CHLNDDEV/OceanMesh2D/files/3907803/TestMergerComplete.zip)

WPringle committed:
fixed for plus with weirs to remove duplicate points

WPringle committed:
Update msh.m

WPringle committed:
fix for bd catch in obj1

WPringle commented:
I will try your tests out, and get back as soon as I can.

WPringle committed:
deleted spurious renumberings for barriers

- barriers contain the height, and coefficients (barincfsp, barincfsp, barinht) which are values not indices so these do not need do be renumbered (caused a segfault when trying to write out the mesh).

krober10nd commented on @msh/msh.m:
```diff
@@ -1572,9 +1576,44 @@ function write(obj,fname,type)
+                % add the deleted elements into the weir triangulation
+                twadd = setdiff(t1,objn.t,'rows');
+                [~,~,I] = intersect(p1(twadd,:),p1o,'rows','stable');
+                twadd = reshape(I,3,3);
```
This line causes a segfault for me because the length of I cannot be reshaped into a 3 by 3 matrix. Is this by accident? I'm trying to understand the logic of this. It seems you remove the weir nodes/elements from the inset mesh and then do the merger. Later on you put the nodes and triangles associated with the layer of weirs back on?

krober10nd commented:
Another problem I've found is that when you have several insets, the typical approach for me is to add them together using minsets=plus(inset1+inset2,inset3,-1) as they are often non-overlapping. This will only carry over the weirs from inset1 and not from inset2 and inset3. When a user goes to add m2=minsets+base ; it will not correctly constrain the weirs because of this.

krober10nd commented:
One solution to that is to make the a method extractWeirs which returns the triangulation and points nearby the weir like such...

        function [pw,tw]=extractWeirs(obj)
            % Return the points and elements associated with the weirs
            weir_nodes = [];
            p1=obj.p; t1=obj.t; 
            for ii = 1:obj.bd.nbou
                if obj.bd.ibtype(ii) ~= 24; continue; end
                nodes = full(obj.bd.nbvv(1:obj.bd.nvell(ii),ii));
                nodes2 = full(obj.bd.ibconn(1:obj.bd.nvell(ii),ii));
                weir_nodes = [weir_nodes; nodes; nodes2];
            end
            % retrieve elements connected to weir nodes
            vtoe = VertToEle(t1);
            connec_eles = vtoe(:,weir_nodes);
            connec_eles(connec_eles == 0) = [];
            connec_eles = unique(connec_eles)';
            tw = t1(connec_eles,:);
            [pw,tw]=fixmesh(p1,tw);
        end

and in msh.plus call 

             % checking for weirs in obj1 to keep
            if ~isempty(obj1.bd) && any(obj1.bd.ibtype == 24)
                disp('Weirs found in obj1, extracting to ensure they are preserved')
                [tw1,pw1]=extractWeirs(obj1); 
            end
            
             % checking for weirs in obj2 to keep
            if ~isempty(obj2.bd) && any(obj2.bd.ibtype == 24)
                disp('Weirs found in obj2, extracting to ensure they are preserved')
                [tw2,pw2]=extractWeirs(obj2); 
            end

krober10nd commented:
Also, I wrote this snippet of code to carry over node strings from both inset1 and inset2 in the merger, which is important for what I said above.  Note this is for the second obj but I just repeated it above it for the first obj to carry over its nodestrings. 


            if ~isempty(obj2.bd) && any(obj2.bd.ibtype == 24)
                disp('Carrying over weir nodestrings from obj2.')
                idx1 = ourKNNsearch(merge.p',obj2.p',1);
                if isempty(merge.bd)
                    merge.bd.nbou=0;
                    merge.bd.nvell=[];
                    merge.bd.nvel=[];
                    merge.bd.nbvv=[];
                    merge.bd.ibconn=[];
                   merge.bd.ibtype=[];
                   merge.bd.barinht=[]; 
                   merge.bd.barincfsb=[];
                   merge.bd.barincfsp=[];
                end
                % number of new boundaries 
                startBou = merge.bd.nbou + 1 ;
                merge.bd.nbou =  merge.bd.nbou + obj2.bd.nbou ;
                % types of boundaries 
                merge.bd.ibtype = [merge.bd.ibtype ; obj2.bd.ibtype]; 
                % new boundaries come after what's already on
                merge.bd.nvell = [merge.bd.nvell; obj2.bd.nvell];
                % nvel is twice the number of nodes on each boundary
                merge.bd.nvel = 2*sum(merge.bd.nvell);
                % nbvv is a matrix of boundary nodes 
                [nr1,nc1]=size(merge.bd.nbvv); 
                [nr2,nc2]=size(obj2.bd.nbvv);
                nbvv_old = merge.bd.nbvv;
                ibconn_old = merge.bd.ibconn;

                merge.bd.nbvv = zeros(max(nr1,nr2),max(nc1,nc1+nc2));
                merge.bd.ibconn = zeros(max(nr1,nr2),max(nc1,nc1+nc2));

                merge.bd.nbvv(1:nr1,1:nc1)=nbvv_old; 
                merge.bd.ibconn(1:nr1,1:nc1)=ibconn_old; 

                % remap
                for ii = startBou:merge.bd.nbou
                
                    idx = ii - (startBou-1) ;
                    nodes =  full(obj2.bd.nbvv(1:obj2.bd.nvell(idx),idx));
                    nodes2 = full(obj2.bd.ibconn(1:obj2.bd.nvell(idx),idx));
  
                    merge.bd.nbvv(1:merge.bd.nvell(ii),ii)   = idx1(nodes);
                    merge.bd.ibconn(1:merge.bd.nvell(ii),ii) = idx1(nodes2);
                end
            end
         
            disp('NB: f15, obj2 f13 data, non-weir nodestrings etc. are not carried over.')                                  
        end

I had success without removing the weirs a priori if the meshes were already cleaned. Note for the meshes I gave you, execute MorichesInlet = clean(MorichesInlet,'passive'); before merging.

Commit:
Building off Will's previous commit. Now able to migrate and lock weirs
from both insets during one merge.
    -If using tight option -1, I shut down projections as the patches
    are very small and the meshes have already considered the projection
    of the earth when they were generated.
    -The msh.plus() could be made more concise by turning some things
    into functions.

Commit:
Small bug fix to ensure ticks are unique before applying to colorbar

Commit:
Adding weir crestline height, super and sub critical coefficients for
weirs

krober10nd commented:
I sent you a test case by email @WPringle ... With that case, I am unable to get the current merging strategy to work. It seems to have some trouble with weirs that are overlapping?

WPringle committed:
additional clean and corrections for msh.plus

- added functions to the merger
- made the tight into a type with string options.
- formalized and sped up the concatenate option, which can be called from plus with the 'match' option
- ported my corrections to the nodal connectivity from the dev branch.
- added option for msh.plot to plot an arbitrary f13 attribute

WPringle committed:
Update msh.m

make default projection equi unless its a global mesh or polar mesh

WPringle committed:
Made updates to collapse_thin_triangles

- For speed I avoid recalculating the gettrimeshquan for whole triangulation and copy it over
- I found that in some cases the pointIDtoreplace would lead to a triangle with the same node number repeated, so I had to put in a catch here, basically this element cannot be deleted.

WPringle committed:
Update msh.m

WPringle committed:
Update msh.m

- fix to be able to detect if unconnected portions can be concatenated instead of going through the merger
- fixing the intersection locking method which did not work for multiple unconnected triangulation

WPringle commented:
I tried you test cases and they work with newest commit.

1. For the three insets (mi, mj, mf) into regional GSBv4 mesh (mo):

- I can do (merging one inset into the updated regional mesh in series):
  mc = (mi + (mj + (mf + mo)));

- And I can do (trivially merging each inset together before merging this combined one into the updated regional mesh):
   mc  = (((mi + mj) + mf) + mo);

- Also, note that the insets have one triangle holes in them that are present before merging. You cannot see these when you plot using the projected triplot because it does not use a background color. 

Example of merged inset into the regional mesh
![image](https://user-images.githubusercontent.com/21131934/70498310-93ea7e80-1adb-11ea-9836-d157cbcbef11.png)

Also note I added a catch to the collapse thin triangles which could lead to an incorrect triangulation in some instances when doing this merging. For some reason this occurrence depended on the projection. 

2. For the latest test case, we have a regional mesh with weirs (mr), and we have an inset (mi) which is just a subset of the regional mesh except inside the subset the weir has been slightly changed. This presents a challenge for the default plus case, but since we know the boundary of the subset matches exactly the nodes from the regional mesh we can use the 'match' type:
   mc = plus(mi,mr,'match'); 
and this works. So the user must be able to tell the merger to use the match case in this situation. I cannot see how we can do it in the other case where there are no matching vertices. It violates the assumptions made.

WPringle committed:
Update msh.m

krober10nd commented:
This looks good to me. I tested it with both cases and it worked great. I like using the terminology of matched and non-matched.  

I think my only comment would be to suppress the text to the screen when using bound_conn_init. Also, when you merge from fort.14's if you locked points for the shoreline then you may lose that in the intersection zone. Oh well.
